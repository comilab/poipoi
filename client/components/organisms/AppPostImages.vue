<template lang="pug">
.flex.flex-col.text-center.relative
  .my-4(
    v-for="(image, i) in post.images",
    :key="image.id",
    v-show="!showAllButtonIsVisible || i === 0"
  )
    AppImageZoomable(
      :src="image.paths.large",
      :large-src="image.paths.original",
      img-class="post-image"
    )

  AppButton.position-x-center.bottom-0.shadow-xl.mb-12.md_mb-16(
    v-if="showAllButtonIsVisible",
    color="white",
    pill,
    :icon="faImages",
    @click="isAllShown = true"
  ) すべて表示 ({{ post.images.length }}枚)
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs, computed } from 'nuxt-composition-api'
import { faImages } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import AppButton from '~/components/atoms/AppButton.vue'
import AppImageZoomable from '~/components/molecules/AppImageZoomable.vue'

export default defineComponent({
  components: {
    AppButton,
    AppImageZoomable
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props) {
    const localState = reactive({
      isAllShown: false
    })

    const showAllButtonIsVisible = computed(() => {
      return props.post.images.length > 1 && !localState.isAllShown
    })

    return {
      ...toRefs(localState),
      showAllButtonIsVisible,
      faImages
    }
  }
})
</script>

<style lang="postcss" scoped>
.post-image {
  max-height: 90vh;
}
</style>
