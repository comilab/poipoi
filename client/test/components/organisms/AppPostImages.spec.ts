import { fireEvent, render } from '@testing-library/vue'

import AppPostImages from '~/components/organisms/AppPostImages.vue'
import imageFactory from '~/test/factories/image'
import postFactory from '~/test/factories/post'

describe('components/organisms/AppPostImages', () => {
  let options: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true
      }
    }
  })

  test('renders correctly', () => {
    options.props.post = postFactory.build({
      images: imageFactory.buildList(2, {
        publicPaths: {
          original: 'original',
          large: 'large',
          medium: 'medium',
          small: 'small'
        }
      })
    })
    const { container } = render(AppPostImages, options)

    expect(container).toMatchSnapshot()
  })

  describe('全て表示', () => {
    beforeEach(() => {
      options.props.post = postFactory.build({
        images: imageFactory.buildList(2)
      })
    })

    test('画像が複数枚ある時のみ表示', async () => {
      options.props.post = postFactory.build({
        images: imageFactory.buildList(1)
      })
      const { queryByRole, updateProps } = render(AppPostImages, options)

      expect(queryByRole('button', { name: /すべて表示/ })).not.toBeInTheDocument()

      await updateProps({
        post: postFactory.build({
          images: imageFactory.buildList(2)
        })
      })
      expect(queryByRole('button', { name: /すべて表示/ })).toBeInTheDocument()
    })

    test('クリックすると画像が全て表示される', async () => {
      const { getAllByRole, getByRole } = render(AppPostImages, options)

      expect(getAllByRole('img')).toHaveLength(1)

      await fireEvent.click(getByRole('button', { name: /すべて表示/ }))
      expect(getAllByRole('img')).toHaveLength(2)
    })
  })
})
