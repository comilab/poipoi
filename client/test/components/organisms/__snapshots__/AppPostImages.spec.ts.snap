// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppPostImages renders correctly 1`] = `
<div>
  <div
    class="flex flex-col text-center relative"
  >
    <div
      class="my-4"
    >
      <div
        class="inline-block"
      >
        <div
          class="zoomable-area relative"
          data-testid="zoomable-area"
        >
          <img
            class="select-none pointer-events-none post-image"
            data-testid="zoomable-area-img"
            src="undefinedlarge"
          />
          <div
            class="caution absolute right-0 bottom-0 text-white text-xs p-2"
          >
            無断転載禁止
          </div>
        </div>
        <transition-stub
          enterclass="opacity-0"
          leavetoclass="opacity-0"
        >
          <!---->
        </transition-stub>
      </div>
    </div>
    <div
      class="my-4"
      style="display: none;"
    >
      <div
        class="inline-block"
      >
        <div
          class="zoomable-area relative"
          data-testid="zoomable-area"
        >
          <img
            class="select-none pointer-events-none post-image"
            data-testid="zoomable-area-img"
            src="undefinedlarge"
          />
          <div
            class="caution absolute right-0 bottom-0 text-white text-xs p-2"
          >
            無断転載禁止
          </div>
        </div>
        <transition-stub
          enterclass="opacity-0"
          leavetoclass="opacity-0"
        >
          <!---->
        </transition-stub>
      </div>
    </div>
    <button
      class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none position-x-center bottom-0 shadow-xl mb-12 md_mb-16 bg-white text-gray-600 hover_bg-gray-300 hover_text-gray-800 px-3 py-2 rounded-full"
      data-testid="button"
      type="button"
    >
      <div>
        <fontawesomeicon-stub
          class="fa-fw mr-2"
          data-testid="icon"
          icon="[object Object]"
        />
      </div>
      <div
        class=""
        data-testid="text"
      >
        すべて表示 (2枚)
      </div>
    </button>
  </div>
</div>
`;
